hack/scripts/kernel.sh

```diff
@@ -0,0 +1,123 @@
+        echo "VMSIZE is empty"
+        flag="true"
+    fi
+    if [ -z $DUMMY_CNI ]; then
+        echo "DUMMY_CNI is empty"
+        flag="true"
+    fi
+    if [ flag = "true" ]; then
+        echo "Please ensure that variables are set correctly."
+        echo "Exiting script gracefully"
+        exit 1
+    fi
+    echo "-- Cluster Create --"
+    make -C ../aks ${CLUSTER_TYPE} \
+    AZCLI=az REGION=${REGION} SUB=${SUB} \
+    CLUSTER=${CLUSTER_NAME} \
+    VM_SIZE=${VMSIZE} \
+    AUTOUPGRADE=none
+else
+    echo Cluster is not being created through script. Proceeding with upgrade.
+fi
+
+if [ $DUMMY_CNI = "true" ]; then
+    echo "-- Install dummy CNI for nodes to be marked as Ready --"
+    kubectl get pods -Aowide
+    kubectl apply -f https://raw.githubusercontent.com/Azure/azure-container-networking/v1.5.3/hack/manifests/cni-installer-v1.yaml
+    kubectl rollout status ds -n kube-system azure-cni
+fi
+
+echo "-- Start privileged daemonset --"
+kubectl get pods -Aowide
+kubectl apply -f ../../test/integration/manifests/load/privileged-daemonset.yaml
+sleep 10s
+kubectl rollout status ds -n kube-system privileged-daemonset
+
+echo "-- Update kernel through daemonset --"
+kubectl get pods -n kube-system -l os=linux,app=privileged-daemonset -owide
+privList=`kubectl get pods -n kube-system -l os=linux,app=privileged-daemonset -owide --no-headers | awk '{print $1}'`
+for pod in $privList; do
+    echo "-- Update Ubuntu Packages --"
+    # Not needed, but ensures that the correct packages exist to perform upgrade
+    kubectl exec -i -n kube-system $pod -- bash -c "apt update && apt-get install software-properties-common -y"
+
+    echo "-- Add proposed repository --"
+    kubectl exec -i -n kube-system $pod -- bash -c "add-apt-repository ppa:canonical-kernel-team/proposed -y"
+    kubectl exec -i -n kube-system $pod -- bash -c "add-apt-repository ppa:canonical-kernel-team/proposed2 -y"
+
+    echo "-- Check apt-cache --"
+    kubectl exec -i -n kube-system $pod -- bash -c "apt-cache madison linux-azure-edge"
+
+    echo "-- Check current Ubuntu kernel --"
+    kubectl exec -i -n kube-system $pod -- bash -c "uname -r"
+    kubectl get node -owide
+
+    echo "-- Install Proposed Kernel --"
+    kubectl exec -i -n kube-system $pod -- bash -c "apt install -y linux-azure-edge"
+done
+
+privArray=(`kubectl get pods -n kube-system -l os=linux,app=privileged-daemonset -owide --no-headers | awk '{print $1}'`)
+nodeArray=(`kubectl get pods -n kube-system -l os=linux,app=privileged-daemonset -owide --no-headers | awk '{print $7}'`)
+kubectl get pods -n kube-system -l os=linux,app=privileged-daemonset -owide
+
+i=0
+for _ in ${privArray[@]}; do
+    echo "-- Restarting Node ${nodeArray[i]} through ${privArray[i]} --"
+    kubectl exec -i -n kube-system ${privArray[i]} -- bash -c "reboot"
+    echo "-- Waiting for condition NotReady --"
+    kubectl wait --for=condition=Ready=false -n kube-system pod/${privArray[i]} --timeout=90s
```

Comment:
how does this work if node has not restarted and pod itself is down ?  Should we check for node `AGE` as well ?

Reply:
So,
1: triggers a reboot for the node on line 106 ` bash -c "reboot"`. This triggers a restart of the node immediately.
2: waits for the priv pod on the matching node to be reconciled and be marked as not ready `--for=condition=Ready=false -n kube-system pod/${privArray[i]} `
3: waits for priv pod to be marked ready `--for=condition=Ready -n kube-system pod/${privArray[i]}`

Not sure if `AGE` will help us here as due to the reconciliation.